File: search/binary-search/src/median.c
/*
 * median.c
 * Copyright (C) 2015 Liu Xinyu (liuxinyu95@gmail.com)
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

/*
 * Find the median of two sorted arrays
 *   O(lg(m+n)) time, and O(1) space
 *
 * Suppose the two arrays are A and B.
 * Perform the following binary search first in A then B to find the median.
 *
 * Start from low = 0, high = |A|, guess i = floor (low + high)/2
 * For the median m, there should be total half = floor (|A| + |B| + 1) / 2 elements not greater than it.
 * Since there are i + 1 elements not greater than A[i] in A,
 * There should be half - (i + 1) elements not greater than A[i] in B.
 * Denote j = half - i - 2, thus we can compare if B[j] <= A[i] <= B[j + 1] is satified. This indicates
 * That the guess is correct median.
 * Otherwise, we can easily tell if the guess is too small or too big, then halve the elements to adjust
 * the guess.
 *
 * [1]. MIT, Introduction to algorithm, problem set 9-1.
 * [2]. Dr. Dobb's Finding the Median of Two Sorted Arrays Efficiently.
 */

#include <stdio.h>
#include <stdlib.h>
#include <math.h>

#define N 1000  /*for verification*/

#define min(x, y) (x < y ? x : y)

int odd(int n) { return n & 0x1; }

void swap(int *x, int *y) {
    int tmp = *x; *x = *y; *y = tmp;
}

/* meidan of an array */
double medianof(int A[], int n) {
    return odd(n) ? (double) A[n / 2] : (double)(A[ n / 2] + A[n / 2 - 1]) / 2.0;
}

int find(int A[], int m, int B[], int n) {
    int l = 0, u = m;
    int i, j, half = (m + n + 1) / 2;
    while (l < u) {
        i = (l + u) / 2;
        j = half - i - 2;
        if (j < 0 || j >= n) {
            if (j == -1 && A[i] <= B[0])
                return i; /* found */
            if (j >= n )
                l = i + 1; /* too small */
            else
                u = i; /* too big */
        } else {
            if (B[j]<= A[i] && (j == n - 1 || A[i] <= B[j+1]))
                return i; /* found */
            else if (A[i] < B[j])
                l = i + 1; /* too small */
            else
                u = i; /* too big */
        }
    }
    return -1;
}

double median(int A[], int m, int B[], int n) {
    int i, j, k, *C;
    if (!A || m == 0)
        return medianof(B, n);
    if (!B || n == 0)
        return medianof(A, m);
    if ((i = find(A, m, B, n)) == -1) {
        i = find(B, n, A, m);
        C = A; A = B; B = C;
        swap(&m, &n);
    }
    if (odd(m + n))
        return (double)A[i];
    j = (m + n) / 2 - i - 2;
    if (i == m - 1)
        k = B[j+1];
    else if (j == n - 1)
        k = A[i+1];
    else
        k = min(A[i+1], B[j+1]);
    return (double)(A[i] + k) / 2.0;
}

/*for verification*/
int cmp(const void* x, const void* y) {
    return *(const int*)x - *(const int*)y;
}

void print(int *xs, int n) {
    int i;
    printf("[");
    for (i = 0; i < n; ++i)
        printf(i == n-1 ? "%d]\n" : "%d, ", xs[i]);
}

void check(int *xs, int n, int* ys, int m, double k) {
    double med = median(xs, n, ys, m);
    if (fabs(med - k) > 0.005) {
        printf("FAIL: expected: %f, actual: %f\n", k, med);
        print(xs, n);
        print(ys, m);
        exit(-1);
    }
    //printf("pass\n\n");
}

int main(int argc, char** argv) {
    int c = N, i, j, n, m;
    int xs[N], ys[N];
    double k;
    while(--c) {
        n = rand() % N + 1;
        for (i = 0; i < n; ++i)
            xs[i] = i;
        k = medianof(xs, n);
        for (i = 0; i < n; ++i) {
            j = rand() % n;
            swap(&xs[i], &xs[j]);
        }
        m = rand() % n;
        for (i = 0; i < m; ++i)
            ys[i] = xs[n - 1 - i];
        n -= m;
        qsort(xs, n, sizeof(int), cmp);
        qsort(ys, m, sizeof(int), cmp);
        check(xs, n, ys, m, k);
    }
    return 0;
}

File: search/binary-search/src/median.py
#
# Find the simplified median for two sorted lists a, b in O(lg(m+n)) time and O(1) space.
# Where m = len(a), n = len(b).
#
# The median is defined (differently from statistics) as:
#   median(x) = x[len(x)/2], where x = merge(a, b)
#
# While in statistics:
#   median(x) = x[n/2] if odd(n) else (x[n/2 - 1] + x[n/2]) / 2.0
#
# Note: we can NOT well define median(x) = m, where m in a + b
#    abs(len([y in a + b, y < m) - len([y in a + b, y > m])) <= 1
# Because the counter example: [0, 1, 2, 3, 3, 3, 3, 3, 5], even if replace < with <=
#

from random import sample, randint

# method 1: binary search based
#
# Perform the following binary search first in a then b:
#
# Start from low = 0, high = |a|, guess i = (low + high) / 2
# For the median m, there should be total half = (|a| + |b|) / 2 elements before it.
# Since there are i elements before a[i] in a, we expect there are j = half - i elements
# before a[i] in b. We check if b[j - 1] <= a[i] <= b[j] holds. If yes, then the guess
# is the median. Otherwise, we adjust low/high to i accordingly to halve the search.
#
# [1]. MIT, Introduction to algorithm, problem set 9-1.
# [2]. Dr. Dobb's Finding the Median of Two Sorted Arrays Efficiently.

def median(a, b):
    if not a:
        return b[len(b) // 2]
    if not b:
        return a[len(a) // 2]
    idx = medof(a, b)
    return median(b, a) if idx is None else a[idx]

def medof(a, b):
    l = 0
    u = len(a)
    while l < u:
        i = (l + u) // 2
        j = (len(a) + len(b)) // 2  - i
        if j < 1 or j >= len(b):
            if (j == 0 and a[i] <= b[0]) or (j == len(b) and b[j - 1] <= a[i]):
                return i    # found
            if j >= len(b):
                l = i + 1   # too small
            else:
                u = i       # too big
        else:
            if b[j - 1]  <= a[i] and a[i] <= b[j]:
                return i    # found
            if a[i] < b[j - 1]:
                l = i + 1   # too small
            else:
                u = i       # too big
    return None

# method 2: use k-th element in a and b.
#
# Suppose m = len(a) >= n = len(b), otherwise, exchange a, b.
# If B == [], return the k-th element of A, if k = 0, return min(a[0], B[0])
# Otherwise, guess j = min(k/2, n), and i = k - j, then compare a[i] and b[j]
# If a[i] < b[j], drop all elements before a[i] and after b[j], then
# recursively find the (k - i)-th element in the rest.
# Othewise, drop all elements before b[j] and after a[i], then recursively
# find the (k-j)-th element in the rest.

def median1(xs, ys):
    n, m = len(xs), len(ys)
    return kth(xs, 0, n, ys, 0, m, (m + n) // 2 + 1)

# find the k-th element from xs[x0, x1) and ys[y0, y1), k starts from 1
def kth(xs, x0, x1, ys, y0, y1, k):
    if x1 - x0 < y1 - y0:
        return kth(ys, y0, y1, xs, x0, x1, k)
    if x1 <= x0:
        return ys[y0 + k - 1]
    if y1 <= y0:
        return xs[x0 + k - 1]
    if k == 1:
        return min(xs[x0], ys[y0])
    j = min(k // 2, y1 - y0)
    i = k - j
    i, j = x0 + i, y0 + j
    if xs[i - 1] < ys[j - 1]:
        return kth(xs, i, x1, ys, y0, j, k - i + x0)
    else:
        return kth(xs, x0, i, ys, j, y1, k - j + y0)

def test(f, n = 100):
    for _ in range(n):
        xs = [randint(0, n // 2) for _ in range(randint(2, n))]
        m = randint(0, len(xs))
        a, b = sorted(xs[:m]), sorted(xs[m:])
        c = sorted(xs)
        med1 = f(a, b)
        med2 = c[len(c) // 2]
        assert med1 == med2, f"err: median({a}, {b}): {med1}, median({c}): {med2}"
    print(f"{n} tests OK")

if __name__ == "__main__":
    test(median)
    test(median1)
